### PR TITLE
CMake: Test Counting

### PR DESCRIPTION
**Pull request by ax3l**
- automatically detect number of examples
- ~~introduce some newlines after column 80~~

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -112,11 +115,21 @@ add_library(example SHARED
   example/example12.cpp
   example/example13.cpp
```
Comment (ax3l) [inline CMakeLists.txt]: just update this list in the future and you are all set

**Comment by wjakob** [comment]
I approve of the ``${PYBIND11_LASTEXAMPLE}`` change but none of the others. The first example variable is always 1 and not needed. Forcing the codebase to 80 lines is IMHO a relic of ancient times, and none of the rest of the code follows this convention (especially not the header files).

**Comment by ax3l** [comment]
`PYBIND11_FIRSTEXAMPLE`: just added that because it is an arbitrary chosen start enumerator (and makes a possible follow-up smaller that might occupy `0` ;) )

lines <= 80 chars: change reverted.
I know it sounds ancient but just keep in mind:
- github adds vertical scrolling (annoying) above ~100-120 chars
- 3-way-diffs for your contributors will soft-wrap or vertically scroll (vimdiff, k3diff, etc.) making it unnecessary hard to solve merge/rebase conflicts
- most distributions open terminals at 80char width

**Comment by ax3l** [comment]
Do you allow me to add some additional comment blocks for structure in a follow-up such as
```CMake
##########################################################
# Project Setup
##########################################################

# cmake cmds with general setup of the project

##########################################################
# Compiler Section
##########################################################

# GCC
# [...] all GCC tweaks her

# MSVC
# [...] all MSVC tweaks

# PGI
# [...] all PGI tweaks

# ...

##########################################################
# Tests
##########################################################

# code a user would not need
```

This would make the file more structured and easier to rip apart (or extend) by users.

**Comment by ax3l** [comment]
note: `master` is currently broken with

```bash
python ./example/example1.py                                                                                                                                     (topic-clean
Traceback (most recent call last):
  File "./example/example1.py", line 6, in <module>
    from example import Example1
ImportError: /home/axel/proggen_partiell/pybind11/example/example.so: undefined symbol: PyInstanceMethod_Type
```

**Comment by wjakob** [comment]
Generally, I think that sort of layout in the cmake file is nice -- however, I doubt that it's possible to structure everything in that way (There are numerous directives before the makefile that must precede the target, and then some more that follow)

**Comment by ax3l** [comment]
all right, if I find the time I will send in a follow up PR on that :)

**Comment by ax3l** [comment]
@wjakob I would put further updates in an upcoming PR. If I did not miss anything and you are ok with it this can be merged.

**Pull request merged**